src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs

```diff
@@ -1728,9 +1728,12 @@ private BoundExpression ReplaceTypeOrValueReceiver(BoundExpression receiver, boo
 
                         foreach (Diagnostic d in typeOrValue.Data.ValueDiagnostics.Diagnostics)
                         {
-                            if (d.Code == (int)ErrorCode.WRN_PrimaryConstructorParameterIsShadowedAndNotPassedToBase &&
+                            // Avoid forcing resolution of lazy diagnostics to avoid cycles.
+                            var code = d is DiagnosticWithInfo { HasLazyInfo: true, LazyInfo.Code: var lazyCode } ? lazyCode : d.Code;
+                            if (code == (int)ErrorCode.WRN_PrimaryConstructorParameterIsShadowedAndNotPassedToBase &&
```

Comment:
Can we instead add e.g. `LazyCode` property to `Diagnostic`? Seems it could be used elsewhere as well, for example:

https://github.com/dotnet/roslyn/blob/462e180642875c0540ae1379e60425f635ec4f78/src/Compilers/CSharp/Portable/Binder/Binder_Crefs.cs#L1007-L1009

Reply:
> Seems it could be used elsewhere as well, for example:

Lazy diagnostics is not necessary limited to Obsolete diagnostics.

Reply:
I'd prefer to keep this as is.